src/QsCompiler/SyntaxProcessor/ExpressionVerification.fs

```diff
@@ -3,801 +3,300 @@
+    exType, intersectDiagnostics @ constrainDiagnostics
+
+/// <summary>
+/// Verifies that <paramref name="expr"/> has a numeric type.
+/// </summary>
+/// <returns>The type of <paramref name="expr"/> and the diagnostics.</returns>
+let private VerifySupportsArithmetic (inference: InferenceContext) expr =
+    expr.ResolvedType, inference.Constrain(expr.ResolvedType, Numeric)
+
+/// <summary>
+/// Verifies that <paramref name="lhs"/> and <paramref name="rhs"/> have an intersecting numeric type.
+/// </summary>
+/// <returns>The intersection type and the diagnostics.</returns>
+let private VerifyArithmeticOp (inference: InferenceContext) range lhs rhs =
+    let exType, intersectDiagnostics = inference.Intersect(lhs.ResolvedType, rhs.ResolvedType)
+    let exType = exType |> ResolvedType.withAllRanges (TypeRange.inferred range)
+    let constrainDiagnostics = inference.Constrain(exType, Numeric)
+    exType, intersectDiagnostics @ constrainDiagnostics
+
+/// <summary>
+/// Verifies that <paramref name="expr"/> has an iterable type.
+/// </summary>
+/// <returns>The iterable item type and the diagnostics.</returns>
+let internal VerifyIsIterable (inference: InferenceContext) expr =
+    let range = rangeOrDefault expr
+    let item = inference.Fresh range
+    item, inference.Constrain(expr.ResolvedType, Iterable item)
+
+/// <summary>
+/// Verifies that <paramref name="lhs"/> and <paramref name="rhs"/> have an intersecting semigroup type.
+/// </summary>
+/// <returns>The intersection type and the diagnostics.</returns>
+let private VerifyConcatenation (inference: InferenceContext) range lhs rhs =
+    let exType, intersectDiagnostics = inference.Intersect(lhs.ResolvedType, rhs.ResolvedType)
+    let exType = exType |> ResolvedType.withAllRanges (TypeRange.inferred range)
+    let constrainDiagnostics = inference.Constrain(exType, Semigroup)
+    exType, intersectDiagnostics @ constrainDiagnostics
+
+/// <summary>
+/// Verifies that <paramref name="lhs"/> and <paramref name="rhs"/> have an intersecting equatable type.
+/// </summary>
+/// <returns>The intersection type and the diagnostics.</returns>
+let private VerifyEqualityComparison (inference: InferenceContext) range lhs rhs =
+    let exType, intersectDiagnostics = inference.Intersect(lhs.ResolvedType, rhs.ResolvedType)
+    let exType = exType |> ResolvedType.withAllRanges (TypeRange.inferred range)
+    let constrainDiagnostics = inference.Constrain(exType, Equatable)
+    intersectDiagnostics @ constrainDiagnostics
+
+/// <summary>
+/// Verifies that <paramref name="exprs"/> can form an array.
+/// </summary>
+/// <returns>The type of the array and the diagnostics.</returns>
+let private VerifyValueArray (inference: InferenceContext) range exprs =
+    let diagnostics = ResizeArray()
+
+    for expr in exprs do
+        if expr.ResolvedType.isMissing then
+            QsCompilerDiagnostic.Error (ErrorCode.MissingExprInArray, []) (rangeOrDefault expr)
```

Comment:
I removed this diagnostic.

The type system does not currently prevent MissingType from appearing in expressions. Syntactically, this is already prevented because `_` is not a valid identifier. Let me know if you also want me to re-add protections at the type system level against MissingType.

Reply:
Come to think of it, how do partial applications work; what exactly determines the type of the resulting callable?

Reply:
Given a partial application expression, `Foo(true, _, (_, 7, _))`, the missing types in the argument:

```qsharp
(Bool, _, (_, Int, _))
```

are instantiated as part of the expected callable type:

```qsharp
(Bool, 'a, ('b, Int, 'c)) -> 'd
```

This is unified with the type of `Foo`. The type of the partial application expression is:

```qsharp
('a, ('b, 'c)) -> 'd
```

Reply:
I found the code in partialArgType. 

Regarding partial applications/correctness of the change:
Another thing that occurred to me and I haven't double checked yet (I am planning to do so as part of the bug bash, but wanted to write it down/mention it now in case I forget): When calling type parameterized callables, do you basically not create new type parameters but instead take the ones that are defined by its signature? Or how else do you enforce that e.g. for function Foo<'A> (a1 : 'A, a2 : 'A) : 'A both arguments are of the same type? Looking at partialArgType, how does that work with partial applications?

Regarding the treatment of MissingType:
If not for partial applications, then MissingType with this change would now basically be another "unit" type (a type with a single value that doesn't support any operators/trait). The special treatment of it in partial applications make the type a bit odd from the type system perspective I think. Form an implementation perspective. I don't see future issues from a user perspective even if the parsing restrictions were lifted, aside from the type being somewhat odd from a language perspective - do you? My judgement here regarding whether it is ok not error for usages of MissingType where it shouldn't be boils down to: Does the implementation still behave reasonably even if we didn't have the restrictions on the parsing side? If the answer is yes, then that makes sense to keep it that way, if the answer is no, then I believe would be better to not rely on the implementation in a different part of the code base.

Reply:
> When calling type parameterized callables, do you basically not create new type parameters but instead take the ones that are defined by its signature? Or how else do you enforce that e.g. for `function Foo<'A> (a1 : 'A, a2 : 'A) : 'A` both arguments are of the same type?

The type parameters are instantiated according to the type parameter list `<'t1, 't2, ... 'tn>`, so `Foo<'A>` gets one instantiated type parameter, e.g. `Foo.'A ↦ 'a0`, resulting in a type of `('a0, 'a0) -> 'a0`.

> Looking at partialArgType, how does that work with partial applications?

Instantiation happens before partial application. When inferring the type of a partial application expression, it would see the already-instantiated type of `Foo`, `('a0, 'a0) -> 'a0`.

> If not for partial applications, then MissingType with this change would now basically be another "unit" type (a type with a single value that doesn't support any operators/trait).

Since missing type unifies with every type, it cannot be the unit type. I believe it would need to be the bottom type to maintain soundness, meaning that if the syntactic restriction were lifted, then the expression `_` would need to abort when evaluated.

Reply:
> Since missing type unifies with every type, it cannot be the unit type. I believe it would need to be the bottom type to maintain soundness, meaning that if the syntactic restriction were lifted, then the expression _ would need to abort when evaluated.

Following this, then `_` would be the expression equivalent of the `fail` statement. So that may or may not be something we want to have but it seems like it would behave reasonably, assuming the runtime follows the right semantics for `_`. (I don't know what happens when C# or QIR generation encounters missing expression.)

Reply:
Right now the runtime doesn't implement that, but relies on missing type not being a valid type (i.e. not occurring) aside from inside partial applications, and type arguments. This change hence basically means that the emission pieces that relied on the parser and type checking before to not permit MissingType except for these few places now rely on the parser to enforce that. All in all, we should see if we can better encapsulate, but for the immediate change I think it largely preserves the status quo. Let's evaluate if it would make sense to eliminate MissingType entirely as a separate work item.